compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -138,8 +138,13 @@ fn univariant_uninterned<'tcx>(
     if optimize {
         let end = if let StructKind::MaybeUnsized = kind { fields.len() - 1 } else { fields.len() };
         let optimizing = &mut inverse_memory_index[..end];
-        let field_align = |f: &TyAndLayout<'_>| {
-            if let Some(pack) = pack { f.align.abi.min(pack) } else { f.align.abi }
+        let effective_field_align = |f: &TyAndLayout<'_>| {
+            if let Some(pack) = pack {
+                f.align.abi.min(pack).bytes()
+            } else {
+                // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
+                f.align.abi.bytes().max(f.size.bytes()).trailing_zeros() as u64
```

Comment:
I don't think the current implementation matches the comment, e.g. for `[u8; 6]`

```
  f.align.abi.bytes().max(f.size.bytes()).trailing_zeros()
= 1                  .max(6             ).trailing_zeros()
= 6                                      .trailing_zeros()
= 0b110                                  .trailing_zeros()
= 1
```

Did you intend something like:
```suggestion
                // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
                f.align.abi.bytes().max(1 << f.size.bytes().trailing_zeros()) as u64
```
which produces `2` for `[u8; 6]`?

Reply:
Okay, yes, it's subtle, but I think you're right.

- because either all fields are packed or not, it's okay for it to sometimes return `align` and sometimes return `log2(align)`, even though that looks wrong
- because size is always a multiple of align, you can't have something like align-4 size-6, which would be computed wrong

Reply:
Actually since you're already relying on the latter property, you could probably just do `f.size.bytes().trailing_zeros()`

Though I'm not sure if that would be better or not

Reply:
I'll add a comment

Reply:
Does the updated comment help?

Reply:
Yeah, looks good